C/Python binding: Add support of `NetworkState::gen_diff()`

cathay4t opened the pull request:
Introducing binding for `NetworkState::gen_diff()`:
 * C: `nmstate_generate_differences()`
 * Python: `libnmstate.generate_differences()`

Unit test case and integration test case included.

liangwen12year commented on rust/src/clib/test/nmstate_gen_diff_test.c:
```diff
@@ -0,0 +1,54 @@
+					 &diff_state,
+					 &err_kind,
+					 &err_msg) == NMSTATE_PASS) {
+		printf("%s\n", diff_state);
```
How do you know that the `diff_state` is correct ? Can we add the string comparison between `diff_state` and `expected_state` ?

cathay4t replied:
This test is just make sure we do not have memory leak in c binding. Integration test case will check whether it works or not.

The yaml output might change indent or other format, so char compare in C is not reliable.

liangwen12year replied:
Got it.

liangwen12year commented on tests/integration/gen_diff_test.py:
```diff
@@ -0,0 +1,56 @@
+def test_gen_diff():
+    des_state = load_yaml(
+        """---
+        interfaces:
```
If `gen_diff` is also applicable to DNS settings (only generate the dns-resolver difference between the current state and desired state)? If not, `pytest --dump-states-with-description-only` will include all the DNS settings in the YAML files.

cathay4t replied:
Yes. The `pytest --dump-states-with-description-only` will only dump diff YAML.

tyll commented on rust/src/python/libnmstate/clib_wrapper.py:
```diff
@@ -200,6 +200,31 @@ def gen_conf(state):
     # pylint: enable=no-member
 
 
+def gen_diff(desired_state, current_state):
```
IMHO it is more typical to have the old (current state) as the first and the new (desired state) as the second argument, similar to the "diff" shell command.

cathay4t replied:
Sure!

cathay4t committed:
C/Python binding: Add support of `NetworkState::gen_diff()`

Introducing binding for `NetworkState::gen_diff()`:
 * C: `nmstate_generate_differences()`
 * Python: `libnmstate.generate_differences()`

Unit test case and integration test case included.

Signed-off-by: Gris Ge <fge@redhat.com>

liangwen12year commented:
@cathay4t  we need to re-vendor the dependencies ?

```CI failure
candidate versions found which didn't match: 3.4.4
location searched: directory source `/usr/share/cargo/registry` (which is replacing registry `crates-io`)
required by package `nmstatectl v2.2.31 (/root/rpmbuild/BUILD/nmstate-2.2.31/rust/src/cli)`
perhaps a crate was updated and forgotten to be re-vendored?
error: Bad exit status from /var/tmp/rpm-tmp.VMiaEH (%build)
```

cathay4t commented:
> @cathay4t we need to re-vendor the dependencies ?
> 
> ```
> candidate versions found which didn't match: 3.4.4
> location searched: directory source `/usr/share/cargo/registry` (which is replacing registry `crates-io`)
> required by package `nmstatectl v2.2.31 (/root/rpmbuild/BUILD/nmstate-2.2.31/rust/src/cli)`
> perhaps a crate was updated and forgotten to be re-vendored?
> error: Bad exit status from /var/tmp/rpm-tmp.VMiaEH (%build)
> ```

No. We have to use older ctrlc in order to compile on RHEL 8.6. This breaks compiling on Fedora.

My plan is to replace `ctrlc` with `tokio/signal`, but that is for another PR.

cathay4t committed:
Merge branch 'base' into gen_diff_clib
